modin/experimental/core/execution/native/implementations/omnisci_on_native/exchange/dataframe_protocol/dataframe.py

```diff
@@ -0,0 +1,332 @@
+        """Raise a ``RuntimeError`` if the way of retrieving the data violates the ``allow_copy`` flag."""
+        if not self._allow_copy and not self._is_zero_copy_possible:
+            raise RuntimeError("Copy required with 'allow_copy=False'")
+
+    __is_zero_copy_possible = None
+
+    @property
+    def _is_zero_copy_possible(self) -> bool:
+        """
+        Check whether it's possible to retrieve data from the DataFrame zero-copy.
+
+        The 'zero-copy' term also means that no extra computations or data transers
+        are needed to access the data.
+
+        Returns
+        -------
+        bool
+        """
+        if self.__is_zero_copy_possible is None:
+            if self._df._has_arrow_table():
+                # If PyArrow table is already materialized then we can
+                # retrieve data zero-copy
+                self.__is_zero_copy_possible = True
+            elif not self._df._can_execute_arrow():
+                # When not able to execute the plan via PyArrow means
+                # that we have to involve OmniSci, so no zero-copy.
+                self.__is_zero_copy_possible = False
+            else:
+                # Check whether the plan for PyArrow can be executed zero-copy
+                self.__is_zero_copy_possible = self._is_zero_copy_arrow_op(self._df._op)
+        return self.__is_zero_copy_possible
+
+    @classmethod
+    def _is_zero_copy_arrow_op(cls, op) -> bool:
+        """
+        Check whether the passed node of the delayed computation tree could be executed zero-copy via PyArrow execution.
+
+        Parameters
+        ----------
+        op : DFAlgNode
+
+        Returns
+        -------
+        bool
+        """
+        is_zero_copy_op = False
+        if isinstance(op, (FrameNode, TransformNode, UnionNode)):
+            # - FrameNode: already materialized PyArrow table
+            # - TransformNode: select certain columns of the table, implemented zero-copy (``df._arrow_select``)
+            # - UnionNode: concatenate PyArrow tables, implemented zero-copy (``df._arrow_concat``)
+            is_zero_copy_op = True
+        elif isinstance(op, MaskNode) and (
+            isinstance(op.row_positions, slice) or is_range_like(op.row_positions)
+        ):
+            # Can select rows zero-copy if indexer is a slice-like (``df._arrow_row_slice``)
+            is_zero_copy_op = True
+        return is_zero_copy_op and all(
+            # Walk the computation tree
+            cls._is_zero_copy_arrow_op(_op)
+            for _op in getattr(op, "inputs", [])
+        )
+
+    @property
+    def _pyarrow_table(self) -> pa.Table:
+        """
+        Get PyArrow table representing the DataFrame.
+
+        Returns
+        -------
+        pyarrow.Table
+        """
+        self._maybe_raise_if_materialize()
+
+        if not self._df._has_arrow_table():
+            self._df._execute()
+
+        at = self._df._partitions[0][0].arrow_table
```

Comment:
`.get()` returns either arrow table or pandas dataframe:
https://github.com/modin-project/modin/blob/cb54844de9e501a2a54c8619a49711477c551b8f/modin/experimental/core/execution/native/implementations/omnisci_on_native/partitioning/partition.py#L95-L97
Accessing an arrow table directly and verifying whether it's None seems more clear rather than `.get()` something from partition and verifying that the return type is correct.

Reply:
ok